File: connectors/java/examples/Quickstart.java
package com.sonnetdb.examples;

import com.sonnetdb.SonnetDbConnection;
import com.sonnetdb.SonnetDbResult;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * SonnetDB Java connector quickstart.
 */
public final class Quickstart {
    private Quickstart() {
    }

    public static void main(String[] args) throws IOException {
        Path dataDir = Files.createTempDirectory("sonnetdb-java-quickstart-");
        run(dataDir);
        System.out.println("data directory: " + dataDir);
    }

    private static void run(Path dataDir) {
        System.out.println("SonnetDB native version: " + SonnetDbConnection.version());

        try (SonnetDbConnection connection = SonnetDbConnection.open(dataDir.toString())) {
            connection.executeNonQuery("CREATE MEASUREMENT cpu (host TAG, usage FIELD FLOAT)");
            int inserted = connection.executeNonQuery(
                "INSERT INTO cpu (time, host, usage) VALUES " +
                    "(1710000000000, 'edge-1', 0.42)," +
                    "(1710000001000, 'edge-1', 0.73)");
            System.out.println("inserted rows: " + inserted);

            try (SonnetDbResult result = connection.execute(
                "SELECT time, host, usage FROM cpu WHERE host = 'edge-1' LIMIT 10")) {
                for (int i = 0; i < result.columnCount(); i++) {
                    if (i > 0) {
                        System.out.print("\t");
                    }
                    System.out.print(result.columnName(i));
                }
                System.out.println();

                while (result.next()) {
                    System.out.printf(
                        "%d\t%s\t%.3f%n",
                        result.getLong(0),
                        result.getString(1),
                        result.getDouble(2));
                }
            }
        }
    }

}

File: connectors/c/examples/quickstart.c
#include <stdio.h>
#include <stdlib.h>

#include "sonnetdb.h"

static void print_last_error(void)
{
    char buffer[1024];
    int32_t written = sonnetdb_last_error(buffer, (int32_t)sizeof(buffer));
    if (written > 0)
    {
        fprintf(stderr, "SonnetDB error: %s\n", buffer);
    }
}

static void require_result(sonnetdb_result* result)
{
    if (result == NULL)
    {
        print_last_error();
        exit(1);
    }
}

int main(void)
{
    sonnetdb_connection* connection = sonnetdb_open("./data-c");
    if (connection == NULL)
    {
        print_last_error();
        return 1;
    }

    sonnetdb_result* result = sonnetdb_execute(
        connection,
        "CREATE MEASUREMENT cpu (host TAG, usage FIELD FLOAT)");
    require_result(result);
    sonnetdb_result_free(result);

    result = sonnetdb_execute(
        connection,
        "INSERT INTO cpu (time, host, usage) VALUES "
        "(1710000000000, 'edge-1', 0.42),"
        "(1710000001000, 'edge-1', 0.73)");
    require_result(result);
    printf("inserted rows: %d\n", sonnetdb_result_records_affected(result));
    sonnetdb_result_free(result);

    result = sonnetdb_execute(
        connection,
        "SELECT time, host, usage FROM cpu WHERE host = 'edge-1' LIMIT 10");
    require_result(result);

    int32_t columns = sonnetdb_result_column_count(result);
    for (int32_t i = 0; i < columns; i++)
    {
        printf("%s%s", i == 0 ? "" : "\t", sonnetdb_result_column_name(result, i));
    }
    printf("\n");

    while (sonnetdb_result_next(result) == 1)
    {
        printf("%lld\t%s\t%.3f\n",
               (long long)sonnetdb_result_value_int64(result, 0),
               sonnetdb_result_value_text(result, 1),
               sonnetdb_result_value_double(result, 2));
    }

    sonnetdb_result_free(result);
    sonnetdb_close(connection);
    return 0;
}
